src/modules/disk_io.cpp

```diff
@@ -0,0 +1,169 @@
+
+  if (!(dp = opendir("/sys/block/")))
+  {
+    m_log.err("Can't open /sys/block!!");
+    return names;
+  }
+  while ((dirp = readdir(dp))) {
+    if (dirp->d_name[0] != '.') {
+      names.push_back(std::string(dirp->d_name));
+    }
+  }
+  closedir(dp);
+  return names;
+}
+
+std::pair<unsigned long long, unsigned long long> disk_io_module::get_disk_read_write(std::string disk_name) {
+  unsigned long long sec_read_total_new{0ULL};
+  unsigned long long sec_write_total_new{0ULL};
+  try {
+    std::ifstream disk_io_info("/sys/block/" + disk_name + "/stat");
+
+    std::string line;
+    std::getline(disk_io_info, line);
+
+    std::vector<std::string> items = string_util::split(line, ' ');
+    std::vector<std::string> filtered_items;
+
+    for (const auto &item : items) {
+      if (item.length() > 0) {
+        filtered_items.push_back(item);
+      }
+    }
+
+    sec_read_total_new = std::strtoull(filtered_items[READ_SECTORS_OFFSET].c_str(), nullptr, 10);
+    sec_write_total_new = std::strtoull(filtered_items[WRITE_SECTORS_OFFSET].c_str(), nullptr, 10);
+  } catch (const std::exception& err) {
+    m_log.err("Failed to read diskstat values (what: %s)", err.what());
+  }
+  return (std::make_pair(sec_read_total_new, sec_write_total_new));
+}
+
+float disk_io_module::get_time_delta(void) {
+  std::chrono::milliseconds now = std::chrono::duration_cast<std::chrono::milliseconds>(
+      std::chrono::system_clock::now().time_since_epoch());
+
+  std::chrono::duration<float> ms_delta = now - m_ms;
+  float time_delta = ms_delta.count();
+  m_ms = now;
+  return time_delta;
+}
+
+void disk_io_module::calculate_disk_io_speeds(std::string disk_name)
+{
+  unsigned long long read_total_new{0ULL};
+  unsigned long long write_total_new{0ULL};
+
+  auto read_write_pair = get_disk_read_write(disk_name);
+  read_total_new = read_write_pair.first;
+  write_total_new = read_write_pair.second;
+  if (read_total_new == 0ULL || write_total_new == 0ULL)
+    return ;
+
+  /*
+   * /sys/block/name/stat third and seventh fields display total completed
+   * reads / writes in standard UNIX blocks (512b) so we need to divide delta
+   * by 2048 to get value in Mb
+   */
+  m_read_speeds[disk_name] = static_cast<float>(read_total_new - m_read_total[disk_name])
+    / (2.0f * m_time_delta);
+  m_write_speeds[disk_name] = static_cast<float>(write_total_new - m_write_total[disk_name])
+    / (2.0f * m_time_delta);
+
+  // replace old values
+  m_read_total[disk_name] = read_total_new;
+  m_write_total[disk_name] = write_total_new;
+}
+
+bool disk_io_module::update() {
+  m_time_delta = get_time_delta();
+
+  float sum_read{0.0f};
+  float sum_write{0.0f};
+
+  for (auto disk_name : m_disk_names) {
+    calculate_disk_io_speeds(disk_name);
+    sum_read += m_read_speeds[disk_name];
+    sum_write += m_write_speeds[disk_name];
+  }
+
+  m_max_read_speed = std::max(m_max_read_speed, sum_read);
+  m_max_write_speed = std::max(m_max_write_speed, sum_write);
```

Comment:
Are the max speeds just the highest speeds measured during the time the module was running?

Reply:
Yes, this method will determine maximum disk read/write speed as soon as read/write to disk will continue longer then module update period. Of course you can perform test write and read from disk for determining this values at initialization of the module, but for me it seems to be a little owerkill for module that will run at long period of time

Reply:
So, if any of the disks don't reach their maximum speeds then the values here are going to be inaccurate?

Reply:
The situation where disks don't reach max speed is highly unlikely and hard to reproduce. But those calculations only made for `<bar-write>` and `<bar-read>`, which are not seem to be very informative at all for displaying precise numbers, they are more suitable for displaying changes in I/O speed, e.g. when you start download directly to hdd, you will see that bar changed dramatically from 0% to 100%, and when download is finished - effect will be opposite.
I've got an idea when I wrote the last sentence: old computers have LED indicators for disk usage, and they (indicators) are quite informative, so you can easily detect if your machine performs some I/O intense actions. Maybe it would be more useful to implement analog for such indicator, using ramp drawtype, as a replacement for bar-read bar-write?